accepted/00000-remote-command-auditing.md

```diff
@@ -0,0 +1,106 @@
+- Salt systems: `path="/manager/systems/cmd"` and `RemoteMinionCommands` Java class.
+
+NOTE: the API call and the `spacecmd` command (`system_runscript`) will still be visible independent of the role.
+
+Additionally, the two mentioned Java classes must implement auditing in the form of logging.
+A new file that will be placed under `/var/log/rhn/` and it will be called `rhn_remote_commands_audit.log` will contain entries in the form:
+
+```
+[timestamp in UTC]
+command executed by the user
+target: 
+  - "local" + system if RemoteCommand on a single system or SSM 
+  - "target-pattern" + pattern if RemoteMinionCommands action
+result
+login of the user in the product that has issued the command
+the IP address of the user originating the command
+```
+
+Example:
+
+```
+[21/Jun/2019:10:57:50 +0000]
+"ls -l"
+"local" + suma-refhead-min-centos7.mgr.suse.de
+"total 0\n-rw-r--r-- 1 mbologna users 0 Jun 21 13:09 test"
+mbologna
+10.31.33.7
+
+[22/Jun/2019:11:32:50 +0000]
+"rm -fr /"
+"target-pattern" + '*'
+""
+nastyuser
+192.168.18.81
+```
+
+The final format of the log could be formatted in a JSON fashion to be easily parseable.
+The log file should be rotated and archived as usual product logs.
+
+The `Remote Command Administrator` role must be explicitly set. By default, the `Remote Command Administrator` role is not assigned to a new user.
+
+Existing users: it has been suggested that existing users should not have the current behavior modified. In that case, the migration script must assign the `Remote Command Administrator` role to all user that have `Administrative Role`s.
+
+## Future Developments
+
+If needed, the `Remote Commands Administrator` role can be customized with a system granularity level: every user can access the `Remote Command` tab on a system if he/she is entitled to do via ACL.
+Additional tables to hold user and systems associated with the role must be introduced and the associated backend code must be implemented.
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+- The file `rhn_remote_commands_audit.log` will contain every command issued by any user (could be a problem in some sensitive environments)
+- Change of expected behavior for new users (change must be documented in release-notes and official documentation)
+
+# Alternatives
+[alternatives]: #alternatives
+
+If we do not implement this feature, there is a lack of security and traceability of every command issues via web or API
+
+# Unresolved questions
+[unresolved]: #unresolved-questions
+
+- Right now, if a user does not have an `Administrative Role`, cannot access any system (he/she does not even see them in the `Systems list`). The new `Remote Command Administrator` should follow the same principle and be an addition to the `Administrative Role` or a new `Administrative Role` by itself?
```

Comment:
Another Question: What makes a `Remote Command Administrator` Role different from the `Configuration Administrator` ?

AFAIK as a `Configuration Administrator` you are allowed to create salt states (and execute them?). You can write `cmd.run` in the state and have the same result as using remote command. This needs to be discussed also with the Champions.

Reply:
If a user only has the `Configuration Administrator` role then you can create states. But you cannot target any system registered to the product (you will have an empty list), neither with `Remote Command` tab (because you don't see any system) neither with `Salt > Remote Commands` with `*` (`no target systems`).
Having a user with just `Remote Command Administrator` is the same as the case above.
By noticing all of the above, the RFC has the unresolved question: is `Remote Command Administrator` something we want in addition to an `Administrative Role` or an `Administrative Role` by itself?

Reply:
Just showed: you can be a normal user and become admin of a system. You do this by adding a system to a system group and make a user "admin" of that system group.

In this case, the user can perform all tasks on that system. Install Software, Call Highstate and run Remote Commands.

Reply:
> AFAIK as a `Configuration Administrator` you are allowed to create salt states (and execute them?). You can write `cmd.run` in the state and have the same result as using remote command. This needs to be discussed also with the Champions.

An user who is not a `Configuration Administrator` cannot apply any state.

Reply:
I think this indicates a deeper problem with the current role structure.

Users are probably using "Administrative" excessively because of some gap in the roles spectrum.

Let's think about it like this:
> If I don't want the user to run remote commands, that means I don't want the user to be an "admin/root". So what other chance do I have (what checkboxes should I check) to have a user with all the permissions equivalent to an admin *minus* remote command execution?

I think the problem here is that only Administrative roles can play around with "all" systems. What's really missing is some collection of roles which would result in a user with access to "all" systems, but with limited capabilities.